hare4/hare_test.go

```diff
@@ -1211,12 +1208,28 @@ func TestHare_ReconstructForward(t *testing.T) {
 					default:
 						panic("bad")
 					}
-					if err := cluster.nodes[other[0]].hare.
-						Handler(ctx, cluster.nodes[i].peerId(), msg); err != nil {
+					if !assert.Eventually(t, func() bool {
+						cluster.nodes[other[0]].hare.mu.Lock()
+						defer cluster.nodes[other[0]].hare.mu.Unlock()
+						_, registered := cluster.nodes[other[0]].hare.sessions[m.Layer]
+						return registered
+					}, 5*time.Second, 50*time.Millisecond) {
+						panic(fmt.Sprintf("node %d did not register in time", other[0]))
+					}
```

Comment:
How about
```suggestion
					require.Eventuallyf(t, func() bool {
						cluster.nodes[other[0]].hare.mu.Lock()
						defer cluster.nodes[other[0]].hare.mu.Unlock()
						_, registered := cluster.nodes[other[0]].hare.sessions[m.Layer]
						return registered
					}, 5*time.Second, 50*time.Millisecond, node %d did not register in time", other[0])
```

Reply:
The reason for the `assert` & `panic` is that if the assertion fails `require` calls `t.Fail` which marks the test as failing and stops the current go routine. This call to `Publish` however is not on the main go routine which is now stuck waiting for it to return.

Normally this can be implemented cleanly with `gomock.WithContext` and passing that context to the code that calls the mocks. I tried updating the code to make that work, but it turned out to be too big of a refactoring to fix a single flaky test.

Instead I cleaned up the code a bit and in the case a `require` fails we just rely on the calling side to time out correctly.